### DISCUSSION TITLE
ItemsSource binding refers to wrong viewmodel in version 11.1

### DISCUSSION THREAD

**Discussion by wgraaf** [AvaloniaUI/Avalonia#16813] [Q&A]
I am populating a sub-menu from an ObservableCollection RecentFiles:

```
    <Menu>
      <MenuItem Header="File">
        <MenuItem Header="Recent" ItemsSource="{Binding RecentFiles}">
          <MenuItem.Styles>
            <Style Selector="MenuItem">
              <Setter Property="Header" Value="{Binding Name}"/>
              <Setter Property="Command" Value="{Binding RecentFileCommand}"/>
            </Style>
          </MenuItem.Styles>
        </MenuItem>
      </MenuItem>
    </Menu>
```

Prior to 11.1 (up to 11.0.13), this worked fine; {Binding Name} and  {Binding RecentFileCommand} resolved to the objects in the RecentFiles ObservableCollection<RecentFileViewModel>.

Starting in 11.1.0, I get the following run time error:

[Binding]An error occurred binding 'Command' to 'RecentFileCommand' at 'RecentFileCommand': 'Could not find a matching property accessor for 'RecentFileCommand' on 'AvTest.ViewModels.**MainWindowViewModel**'

Apparently, the binding is attempting resolution on  MainWindowViewModel, which holds the ObservableCollection; not to elements in  RecentFiles ObservableCollection<**RecentFileViewModel**>.

Below is  the viewmodel code:

```
  public class MainWindowViewModel : ViewModelBase
  {
    public ObservableCollection<RecentFileViewModel> RecentFiles { get; } = new();

    public MainWindowViewModel()
    {
      RecentFiles.Add(new RecentFileViewModel("file1"));
      RecentFiles.Add(new RecentFileViewModel("file2"));
      RecentFiles.Add(new RecentFileViewModel("file3"));
    }
  }

  public class RecentFileViewModel
  {
    public string Name { get; }
    public ReactiveCommand<Unit, Unit> RecentFileCommand { get; }

    public RecentFileViewModel(string name)
    {
      Name = name;
      RecentFileCommand = ReactiveCommand.Create(RecentFile);
    }

    private void RecentFile() => Trace.WriteLine($"Recent File:{Name}");
  }
```

Is a different markup needed to address this issue in 11.1?

Attempting compiled bindings results in slightly different runtine error:
[Binding]An error occurred binding 'Command' to 'RecentFileCommand' at 'RecentFileCommand': 'Unable to cast object of type 'AvTest.ViewModels.MainWindowViewModel' to type 'AvTest.ViewModels.RecentFileViewModel'

**Comment by stevemonaco** [comment]
I try to avoid inline styles, especially in this scenario. I use this approach:

```xml
<UserControl.Styles>
    <Style Selector="MenuItem.recent > MenuItem">
        <Setter Property="Header" Value="{Binding}" />
        <Setter Property="Command" Value="{Binding #menu.((vm:MenuViewModel)DataContext).OpenRecentProjectCommand}" />
        <Setter Property="CommandParameter" Value="{Binding}" />
    </Style>
</UserControl.Styles>
```

```xml
<MenuItem Classes="recent" Header="Open Recent Project" IsVisible="{Binding RecentProjectFiles.Count}" ItemsSource="{Binding RecentProjectFiles}" />
```

I'm binding to an `ObservableCollection<string>` here, but it should work with a concrete VM type. Also, my `Command` is on the MenuViewModel, so you can simplify a bit on your end.

**Reply by wgraaf** [reply]
Thanks for the reply; I have not gotten that approach to work yet.

Still, it seems that the approach with the inline style is broken: starting with version 11.1 to binding is attempting to bind to **MainWindowViewModel**, NOT to the collection element **RecentFileViewModel**

**Reply by stevemonaco** [reply]
You can check out https://github.com/AvaloniaUI/Avalonia/issues/9372 and https://github.com/AvaloniaUI/Avalonia/issues/3100#issuecomment-753452752

Basically, inline style that only selects for "MenuItem" will apply to the MenuItem itself (eg. the `<MenuItem Header="Recent" ItemsSource="{Binding RecentFiles}">`) as well as its children.

**Reply by wgraaf** [reply]
The command binding kept giving me an exception: System.ArgumentException: 'Unable to resolve type vm:MainWindowViewModel from any of the following locations: '

I got it working (both with app- and inline style) by referencing the parent as follows:

```
      <Style Selector="MenuItem.recent > MenuItem">
        <Setter Property="Header" Value="{Binding}" />
        <Setter Property="Command" Value="{Binding $parent.Command}" />
        <Setter Property="CommandParameter" Value="{Binding}" />
      </Style>

```
and
`        <MenuItem Header="Recent" Classes="recent" ItemsSource="{Binding RecentFileNames}" Command="{Binding OpenRecentFileCommand}"/>
`
